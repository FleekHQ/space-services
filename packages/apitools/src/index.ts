import { APIGatewayProxyResult } from 'aws-lambda';

interface ProcessRequestOpts {
  successCode?: number;
  errorCode?: number;
}

interface ErrorToStatusCodeInterface {
  [key: string]: number;
}

const mapErrorToStatusCode: ErrorToStatusCodeInterface = {
  ValidationError: 422,
  NotFoundError: 404,
  Error: 500,
};

const getErrorMessage = (err: Error): string => {
  // If the error is a validation error, we return its message
  if (err.name === 'ValidationError' || err.name === 'NotFoundError') {
    return err.message;
  }

  // We obfuscate it otherwise
  return 'Unexpected server error';
};

const DEFAULT_ERROR_CODE = 500;
const DEFAULT_SUCCESS_CODE = 200;

export const renderError = (
  err: Error,
  statusCode: number | undefined
): APIGatewayProxyResult => {
  let sc = statusCode;

  if (!sc) {
    sc = mapErrorToStatusCode[err.name] || DEFAULT_ERROR_CODE;
  }
  return {
    statusCode: sc,
    body: JSON.stringify(
      {
        message: getErrorMessage(err),
      },
      null,
      2
    ),
  };
};

export const renderSuccess = (
  response: object,
  statusCode: number = DEFAULT_SUCCESS_CODE
): APIGatewayProxyResult => {
  return {
    statusCode,
    body: JSON.stringify(response, null, 2),
  };
};

export const processRequest = async (
  requestFn: Function,
  opts: ProcessRequestOpts = {}
): Promise<APIGatewayProxyResult> => {
  let result;
  try {
    result = await requestFn();
  } catch (err) {
    console.error('Error when processing request: ', err.message);
    console.error(err.stack);
    return renderError(err, opts.errorCode);
  }

  return renderSuccess(result, opts.successCode);
};
