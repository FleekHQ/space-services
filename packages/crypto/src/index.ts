/* eslint-disable import/prefer-default-export */
import { SHA3 } from 'sha3';

/**
 * Obtains an address given a public key
 * @param pubKey The public key (in hex encoding)
 */
export const deriveAddressFromPubKey = (pubKey: string): string => {
  const hash = new SHA3(256);

  // Compute the SHA3-256 hash of the public key
  hash.update(pubKey, 'hex');

  // Get the hex representation of the SHA3-256 hash
  const hexHash = hash.digest('hex');

  // Drop the first 14 bytes (28 characters)
  const trimmedHash = hexHash.substring(28);

  return `0x${trimmedHash}`;
};

export const isValidAddress = (address: string): boolean => {
  const regex = /^0x[0-9A-Fa-f]{36}$/;
  if (address.match(regex)) {
    return true;
  }

  return false;
};
