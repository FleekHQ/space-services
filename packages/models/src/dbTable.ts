import { DocumentClient } from 'aws-sdk/clients/dynamodb';
import dynamodbUpdateExpression from 'dynamodb-update-expression';

class DbTable {
  client!: DocumentClient;

  table: string;

  constructor(
    table: string,
    client: DocumentClient = new DocumentClient({ region: 'us-west-2' })
  ) {
    this.table = table;
    this.client = client;
  }

  put(
    data: DocumentClient.PutItemInputAttributeMap
  ): Promise<DocumentClient.PutItemOutput> {
    const params: DocumentClient.PutItemInput = {
      TableName: this.table,
      Item: data,
    };
    return this.client.put(params).promise();
  }

  // another get with different arguments for convenience
  // Keys are handled as pk and sk
  getItem(
    partitionKey: string,
    sortKey: string
  ): Promise<DocumentClient.GetItemOutput> {
    const params = {
      TableName: this.table,
      Key: {
        pk: partitionKey,
        sk: sortKey,
      },
    };
    return this.client.get(params).promise();
  }

  // another update for convenience
  async updateItem(
    partitionKey: string,
    sortKey: string,
    updateData: any,
    originalItem?: DocumentClient.GetItemOutput
  ): Promise<DocumentClient.UpdateItemOutput> {
    let original: DocumentClient.GetItemOutput;
    if (originalItem) {
      original = originalItem;
    } else {
      original = await this.getItem(partitionKey, sortKey);
    }

    const updateExpression = dynamodbUpdateExpression.getUpdateExpression(
      original,
      updateData
    );

    const params = {
      TableName: this.table,
      Key: {
        pk: partitionKey,
        sk: sortKey,
      },
      ReturnValues: 'ALL_NEW',
      ConditionExpression: 'attribute_exists(pk) AND attribute_exists(sk)',
      ...updateExpression,
    };

    if (params.UpdateExpression === '') {
      return Promise.resolve({});
    }

    return new Promise((resolve, reject) => {
      return this.client
        .update(params)
        .promise()
        .then(result => {
          if (result.Attributes.data) {
            resolve(result.Attributes.data);
          } else {
            resolve(result.Attributes);
          }
        })
        .catch(reject);
    });
  }

  // another put for convenience
  putItem(
    partitionKey: string,
    sortKey: string,
    data: any
  ): Promise<DocumentClient.PutItemOutput> {
    // NOTE: ConditionExpression avoids adding with put an item with duplicate keys
    const params = {
      TableName: this.table,
      Item: {
        pk: partitionKey,
        sk: sortKey,
        data,
      },
      ConditionExpression:
        'attribute_not_exists(pk) AND attribute_not_exists(sk)',
    };

    return this.client.put(params).promise();
  }

  // convencience method to upsert and get back updated item
  // Idea taken from http://www.jramoyo.com/2017/03/upserting-items-into-dynamodb.html
  async upsertItem(
    partitionKey: string,
    sortKey: string,
    data: any
  ): Promise<DocumentClient.GetItemOutput> {
    // 1. get existing item
    const existingItem = await this.getItem(partitionKey, sortKey);
    if (Object.keys(existingItem).length > 0) {
      // 2. update item if it already exists
      await this.updateItem(partitionKey, sortKey, data, existingItem);
    } else {
      // 3. Otherwise, put/insert the item
      try {
        await this.putItem(partitionKey, sortKey, data);
      } catch (err) {
        if (err.code === 'ConditionalCheckFailedException') {
          console.log(
            `upsert ConditionalCheckFailedException pk: ${partitionKey} sk: ${sortKey}`
          );
          // 3a. Only 1 of the concurrent puts will succeed,
          // the rest should retry recursively
          // NOTE: @doug this code is only for race conditions in case
          // multiples put with same keys happen which is unlikely
          // leaving it commented
          // return this.upsertItem(partitionKey, sortKey, data);
        }
        throw err;
      }
    }
    // call item again after update
    return this.getItem(partitionKey, sortKey);
  }

  get(data: DocumentClient.Key): Promise<DocumentClient.GetItemOutput> {
    const params = {
      TableName: this.table,
      Key: data,
    };
    return this.client.get(params).promise();
  }

  query(data: DocumentClient.QueryInput): Promise<DocumentClient.QueryOutput> {
    return this.client.query(data).promise();
  }

  update(
    data: DocumentClient.UpdateItemInput
  ): Promise<DocumentClient.UpdateItemOutput> {
    return this.client.update(data).promise();
  }

  delete(data: DocumentClient.Key): Promise<DocumentClient.DeleteItemOutput> {
    const params = {
      TableName: this.table,
      Key: data,
    };

    return this.client.delete(params).promise();
  }

  /**
   * Writing and reading in batches
   * https://docs.amazonaws.cn/en_us/sdk-for-javascript/v2/developer-guide/dynamodb-example-table-read-write-batch.html
   * @param requests
   */

  batchWrite(
    requests: DocumentClient.WriteRequests
  ): Promise<DocumentClient.BatchWriteItemOutput> {
    const params = {
      RequestItems: {
        [this.table]: requests,
      },
    };
    return this.client.batchWrite(params).promise();
  }

  batchGet(
    requests: DocumentClient.KeysAndAttributes
  ): Promise<DocumentClient.BatchGetItemOutput> {
    const params = {
      RequestItems: {
        [this.table]: requests,
      },
    };
    return this.client.batchGet(params).promise();
  }

  async getWithError(
    data: DocumentClient.Key,
    notFoundMessage?: string
  ): Promise<DocumentClient.AttributeMap> {
    const result = await this.get(data);

    if (!result || !result.Item) {
      throw new Error(notFoundMessage || 'Record was not found.');
    }

    return result.Item;
  }
}

export default DbTable;
