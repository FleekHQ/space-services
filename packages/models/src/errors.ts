/* eslint-disable max-classes-per-file */

export class DomainError extends Error {
  constructor(message) {
    super(message);
    // Ensure the name of this error is the same as the class name
    this.name = this.constructor.name;
    // This clips the constructor invocation from the stack trace.
    // It's not absolutely essential, but it does make the stack trace a little nicer.
    //  @see Node.js reference (bottom)
    Error.captureStackTrace(this, this.constructor);
  }
}

export class ValidationError extends DomainError {
  constructor(message) {
    super(`Validation Error: ${message}`);
  }
}

export class NotFoundError extends DomainError {
  constructor(message) {
    super(`Not Found Error: ${message}`);
  }
}
