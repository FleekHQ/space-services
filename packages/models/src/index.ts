export { SignatureModel } from './signatures';
export { IdentityModel } from './identity';
export { VaultModel } from './vault';
export { default as createNotificationsModel } from './notifications';
export { AccountWithBilling, default as createBillingModel } from './billing';
export * from './errors';
export default {};
