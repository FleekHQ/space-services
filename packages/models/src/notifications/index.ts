import DbTable from '../dbTable';

type Connection = {
  pk: string;
  sk: string;
  connectionId: string;
  establishedAt: string;
  gs1pk: string;
  gs1sk: string;
};

interface ModelApi {
  getConnectionByUuid: (uuid: string) => Promise<Connection>;
  saveConnection: (uuid: string, connectionId: string) => Promise<Connection>;
  deleteConnectionById: (connectionId: string) => Promise<void>;
}

const SORT_KEY = 'wss';

export default (env: string): ModelApi => {
  const db = new DbTable(`space_table_${env}`);

  return {
    getConnectionByUuid: async (uuid: string): Promise<Connection> => {
      const KeyConditionExpression = 'gs1pk = :id AND begins_with(gs1sk, :sk)';
      const ExpressionAttributeValues = {
        ':id': uuid,
        ':sk': SORT_KEY,
      };

      const params = {
        TableName: db.table,
        Select: 'ALL_PROJECTED_ATTRIBUTES',
        KeyConditionExpression,
        ExpressionAttributeValues,
        IndexName: 'gs1',
        ScanIndexForward: false,
      };

      const output = await db.query(params);

      if (!output.Items.length) {
        throw new Error(
          `Websocket Connection for user "${uuid}" was not found.`
        );
      }

      const lastConnection = output.Items.shift() as Connection;

      return lastConnection;
    },

    saveConnection: async (
      uuid: string,
      connectionId: string
    ): Promise<Connection> => {
      const establishedAt = new Date().toISOString();

      const rawInfo = {
        pk: connectionId,
        sk: SORT_KEY,
        gs1pk: uuid,
        gs1sk: `${SORT_KEY}#${establishedAt}`,
        connectionId,
        establishedAt,
      };
      await db.put(rawInfo);

      return rawInfo;
    },

    deleteConnectionById: async (connectionId: string): Promise<void> => {
      await db.delete({
        pk: connectionId,
        sk: SORT_KEY,
      });
    },
  };
};
