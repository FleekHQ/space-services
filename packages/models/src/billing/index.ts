import { DocumentClient } from 'aws-sdk/clients/dynamodb';
import { Wallet, RawWallet } from './types';
import base58Keygen from './base58-keygen';
import { BaseModel } from '../base';
import { PrimaryKey } from '../types';

import { NotFoundError, ValidationError } from '../errors';

const WALLET_KEY = 'wallet';

export const getWalletPrimaryKey = (key: string): PrimaryKey => ({
  pk: key,
  sk: WALLET_KEY,
});

const mapWalletToDbObject = (wallet: Wallet): RawWallet => {
  const { ownerUuid, key, ...rest } = wallet;

  return {
    ...getWalletPrimaryKey(key),
    gs1pk: ownerUuid,
    gs1sk: WALLET_KEY,
    ...rest,
  };
};

const parseDbObjectToWallet = (raw: RawWallet): Wallet => {
  const wallet: Wallet = {
    key: raw.pk,
    ownerUuid: raw.gs1pk,
    credits: raw.credits,
    createdAt: raw.createdAt,
  };

  return wallet;
};

export class WalletModel extends BaseModel {
  constructor(env: string, client: DocumentClient = new DocumentClient()) {
    const table = `space_table_${env}`;
    super(table, client);
  }

  public async createWallet(): Promise<Wallet> {
    const createdAt = new Date(Date.now()).toISOString();
    const key = base58Keygen(16);

    let existingWallet = null;
    try {
      existingWallet = this.getWalletByKey(key);
    } catch (error) {
      // Do nothing
    }

    if (existingWallet != null) {
      // Very unlikely error (1 in 1.64 * 10^28)
      throw new ValidationError('Wallet key collision. Please try again.');
    }

    const newWallet: Wallet = {
      createdAt,
      ownerUuid: '',
      key,
      credits: 0,
    };

    const dbItem = mapWalletToDbObject(newWallet);

    await this.put(dbItem);

    return newWallet;
  }

  public async getWalletByKey(key: string): Promise<Wallet> {
    const rawWallet = await this.get(getWalletPrimaryKey(key)).then(
      result => result.Item as RawWallet
    );

    if (!rawWallet) {
      throw new NotFoundError(`Wallet with key ${key} not found.`);
    }

    return parseDbObjectToWallet(rawWallet);
  }
}

export default WalletModel;
