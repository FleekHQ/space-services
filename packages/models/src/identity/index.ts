import { DocumentClient } from 'aws-sdk/clients/dynamodb';
import { deriveAddressFromPubKey } from '@packages/crypto';
import { v4 as uuidv4 } from 'uuid';
import {
  mapIdentityDbObject,
  mapProofDbObject,
  parseDbObjectToIdentity,
  mapUsernameDbObject,
  mapAddressDbObject,
  getAddressPrimaryKey,
  parseDbObjectToAddress,
  getIdentityPrimaryKey,
  getUsernamePrimaryKey,
  parseDbObjectToUsername,
} from './access-patterns';
import {
  CreateIdentityInput,
  IdentityRecord,
  CreateProofInput,
  ProofRecord,
  RawIdentityRecord,
  RawAddressRecord,
} from './types';
import { validateIdentity } from './validations';
import { BaseModel } from '../base';
import { NotFoundError, ValidationError } from '../errors';

export class IdentityModel extends BaseModel {
  constructor(env: string, client: DocumentClient = new DocumentClient()) {
    const table = `space_table_${env}`;
    super(table, client);
  }

  public async createIdentity(
    input: CreateIdentityInput
  ): Promise<IdentityRecord> {
    const createdAt = new Date(Date.now()).toISOString();
    const address = deriveAddressFromPubKey(input.publicKey);

    const newIdentity = {
      uuid: uuidv4(),
      address,
      createdAt,
      publicKey: input.publicKey,
      username: input.username,
    };

    await validateIdentity(this, newIdentity);

    const dbItem = mapIdentityDbObject(newIdentity);

    // reserve username for this identity
    if (newIdentity.username) {
      await this.put(mapUsernameDbObject(newIdentity));
    }
    // reserve address for this identity
    await this.put(mapAddressDbObject(newIdentity));

    // create identity
    await this.put(dbItem);

    return newIdentity;
  }

  public async createProof(input: CreateProofInput): Promise<ProofRecord> {
    const createdAt = new Date(Date.now()).toISOString();

    const newProof = {
      proofType: input.proofType,
      uuid: input.uuid,
      value: input.value,
      createdAt,
    };

    const dbItem = mapProofDbObject(newProof);

    await this.put(dbItem);

    return newProof;
  }

  public async getIdentityByUuid(uuid: string): Promise<IdentityRecord> {
    const rawIdentity = await this.get(getIdentityPrimaryKey(uuid)).then(
      result => result.Item as RawIdentityRecord
    );

    if (!rawIdentity) {
      throw new NotFoundError(`Identity was not found.`);
    }

    return parseDbObjectToIdentity(rawIdentity);
  }

  public async getIdentityByAddress(address: string): Promise<IdentityRecord> {
    const rawAddress = await this.get(getAddressPrimaryKey(address)).then(
      result => result.Item as RawAddressRecord
    );

    if (!rawAddress) {
      throw new NotFoundError(`Identity with address ${address} not found.`);
    }

    const record = parseDbObjectToAddress(rawAddress);
    return this.getIdentityByUuid(record.uuid);
  }

  public getIdentityByPublicKey(publicKey: string): Promise<IdentityRecord> {
    const address = deriveAddressFromPubKey(publicKey);
    return this.getIdentityByAddress(address);
  }

  public async getIdentityByUsername(
    username: string
  ): Promise<IdentityRecord> {
    const rawUsername = await this.get(getUsernamePrimaryKey(username)).then(
      result => result.Item as RawAddressRecord
    );

    if (!rawUsername) {
      throw new NotFoundError(`Identity with username ${username} not found.`);
    }

    const record = parseDbObjectToUsername(rawUsername);
    return this.getIdentityByUuid(record.uuid);
  }

  public async updateIdentity(uuid: string, payload: Record<string, any>) {
    // check that identity exists
    await this.getIdentityByUuid(uuid);

    const allowedKeys = ['displayName', 'avatarUrl'];

    // update uuid with new username
    const Key = getIdentityPrimaryKey(uuid);

    const ExpressionAttributeValues = {};
    const ExpressionAttributeNames = {};

    const updates = [];

    Object.keys(payload).forEach(key => {
      if (allowedKeys.indexOf(key) >= 0) {
        ExpressionAttributeValues[`:${key}`] = payload[key];
        ExpressionAttributeNames[`#${key}`] = key;
        updates.push(`#${key} = :${key}`);
      }
    });

    if (!updates.length) {
      throw new Error('Invalid payload.');
    }

    const UpdateExpression = `SET ${updates.join(',')}`;

    const params = {
      TableName: this.table,
      Select: 'ALL_PROJECTED_ATTRIBUTES',
      Key,
      ExpressionAttributeValues,
      ExpressionAttributeNames,
      UpdateExpression,
    };

    await this.update(params);
  }

  public async changeUsername(uuid: string, username: string) {
    const usernameExists = await this.getIdentityByUsername(username).catch(
      () => false
    );

    if (usernameExists) {
      throw new ValidationError('Username is already taken.');
    }

    const identity = await this.getIdentityByUuid(uuid);

    // assign new username to uuid
    await this.put(
      mapUsernameDbObject({
        uuid,
        username,
        createdAt: new Date().toISOString(),
      })
    );

    // update uuid with new username
    const Key = getIdentityPrimaryKey(uuid);

    const ExpressionAttributeValues = {
      ':username': username,
    };
    const ExpressionAttributeNames = {
      '#username': 'username',
    };
    const UpdateExpression = 'SET #username = :username';

    const params = {
      TableName: this.table,
      Select: 'ALL_PROJECTED_ATTRIBUTES',
      Key,
      ExpressionAttributeValues,
      ExpressionAttributeNames,
      UpdateExpression,
    };

    await this.update(params);

    // release old username
    await this.delete(getUsernamePrimaryKey(identity.username));
  }
}

export default IdentityModel;
