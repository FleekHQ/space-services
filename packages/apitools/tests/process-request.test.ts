import { processRequest } from '../src';

describe('processRequest', () => {
  it('returns successfully if process does not fail', () => {
    const result = processRequest(() => {
      return {
        key: 'value',
      };
    });

    expect(result).toMatchSnapshot();
  });

  it('returns error message if the process fails', () => {
    const result = processRequest(() => {
      throw new Error('Some error');
    });

    expect(result).toMatchSnapshot();
  });

  it('works for async processes', () => {
    const result = processRequest(() => {
      return new Promise(resolve => {
        setTimeout(() => {
          resolve({ result: 'Async invocation' });
        }, 10);
      });
    });

    expect(result).toMatchSnapshot();
  });
});
