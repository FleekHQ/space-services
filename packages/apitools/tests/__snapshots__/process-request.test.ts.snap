// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processRequest returns error message if the process fails 1`] = `
Object {
  "body": "{
  \\"message\\": \\"Some error\\"
}",
  "statusCode": 500,
}
`;

exports[`processRequest returns successfully if process does not fail 1`] = `
Object {
  "body": "{
  \\"key\\": \\"value\\"
}",
  "statusCode": 200,
}
`;

exports[`processRequest works for async processes 1`] = `
Object {
  "body": "{}",
  "statusCode": 200,
}
`;
